Source/JavaScriptCore/parser/Parser.cpp

```diff
@@ -5024,6 +5026,60 @@ template <class TreeBuilder> TreeExpression Parser<LexerType>::createResolveAndU
     return context.createResolve(location, *ident, start, lastTokenEndPosition());
 }
 
+template <typename LexerType>
+template <class TreeBuilder> TreeExpression Parser<LexerType>::tryParseArgumentsDotLengthForFastPath(TreeBuilder& context, const JSTokenLocation& startLocation, const JSTextPosition& expressionStart)
+{
+    // There is a fast path for getting `arguments.length` by reading `argumentCountIncludingThis`
+    // directly from CallFrame. In that case, no need to materialize `arguments` object. The fast
+    // path for `arguments.length` is applied by excluding 'arguments.length` pattern for
+    // ArgumentsFeature except for two cases:
+    // 1. 'arguments.length` modifications.
+    // 2. Function level global variable declaration with identifier `arguments`.
+    if (context.hasArgumentsFeature() || !match(IDENT) || !isArgumentsIdentifier())
+        return 0;
+
+    // If semantic checks fail here, then let `parsePrimaryExpression` handle the error thrown.
+    // Note that these checks must align to the checks in `parsePrimaryExpression` under
+    // the clause with token type IDENT.
+    if (UNLIKELY(currentScope()->isStaticBlock()
+        || m_parserState.isParsingClassFieldInitializer
+        || currentScope()->evalContextType() == EvalContextType::InstanceFieldEvalContext))
+        return 0;
+
+    SavePoint argumentsSavePoint = createSavePoint(context);
+    JSTextPosition primaryStart = tokenStartPosition();
+    JSTokenLocation primaryLocation(tokenLocation());
+    next();
+    if (match(DOT)) {
+        JSTextPosition expressionEnd = lastTokenEndPosition();
+        next();
+        if (match(IDENT) && *m_token.m_data.ident == m_vm.propertyNames->length) {
+            m_seenArgumentsDotLength = true;
+
+            // Patterns `arguments.length.p1` and `arguments.length["p1"]` are not allowed.
+            SavePoint argumentsDotLengthSavePoint = createSavePoint(context);
+            next();
+            if (match(DOT) || match(OPENBRACKET))
+                goto failToFindArgumentsDotLengthPattern;
+            restoreSavePoint(context, argumentsDotLengthSavePoint);
```

Comment:
Why? I think what we should do is just creating an argument ResolveNode without marking usesArguments. That's it.